### DISCUSSION TITLE
Using records in .NET Framework: possible risks?

### DISCUSSION THREAD

**Discussion by vuplea** [dotnet/runtime#72922] [Q&A] [answered]
You can apparently use records in .NET Framework if you define `System.Runtime.CompilerServices.IsExternalInit` as [such](https://github.com/dotnet/runtime/issues/43638#issue-725644764).

Are there any foreseeable risks in doing this (and using records)?

**Comment by vuplea** [comment]
This [article](https://btburnett.com/csharp/2020/12/11/csharp-9-records-and-init-only-setters-without-dotnet5.html) by Brant Burnett suggests it should be fine for internal use.

**Comment by huoyaoyuan** [comment, marked as answer]
Record does not rely on any runtime feature support, so there should not be such risks.
The risks of using record don't apply to .NET Framework only. Old compilers/serializers may not support `init` setters properly.

**Reply by danmoseley** [reply]
cc @jaredpar

**Reply by tannergooding** [reply]
Records are largely "just syntax sugar" and don't rely on any runtime specific behavior so they should work downlevel on older runtimes such as .NET Framework.

However, newer versions of C# are officially unsupported on older runtimes as documented in https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version and in many other locations.
> [C# 11](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-11) is supported only on .NET 7 and newer versions. [C# 10](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-10) is supported only on .NET 6 and newer versions. [C# 9](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-9) is supported only on .NET 5 and newer versions. [C# 8.0](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8) is supported only on .NET Core 3.x and newer versions.

When using a new language feature in an unsupported scenario the exact behavior you may get could differ and the overall experience around using the feature may end up feeling incomplete. For .NET Framework and .NET Standard, you should be targeting C# 7.3 as that is the highest officially supported language version.

You can of course, at your own discretion, target a higher language version and there are a few projects that do this successfully. It is just important to keep in mind that it is unsupported and so bugs or other issues that crop up (especially if they only impact the downlevel runtime) may be closed as "by design, this is an unsupported scenario".
